src/f32-gemm/MrxLmulv-rvv.c.in

```diff
@@ -0,0 +1,170 @@
+// Copyright 2023 SiFive, Inc.
+//
+// This source code is licensed under the BSD-style license found in the
+// LICENSE file in the root directory of this source tree.
+
+$assert DATATYPE in ["F32"]
+$assert ACTIVATION in ["LINEAR", "RELU", "MINMAX"]
+$assert MR >= 1
+$assert LMUL in [1, 2, 4]
+#include <assert.h>
+
+#include <riscv_vector.h>
+
+#include <xnnpack/gemm.h>
+
+
+$DATATYPE_SPEC = {"F32": "f32"}[DATATYPE]
+$SUFFIX = {"LINEAR": "", "RELU": "_relu", "MINMAX": "_minmax"}[ACTIVATION]
+$PARAMS = {"LINEAR": "xnn_f32_default_params", "RELU": "xnn_f32_relu_params", "MINMAX": "xnn_f32_minmax_params"}[ACTIVATION]
+void xnn_${DATATYPE_SPEC}_gemm${SUFFIX}_ukernel_${MR}x${LMUL}v__rvv(
+    size_t mr,
+    size_t nc,
+    size_t kc,
+    const float* restrict a,
+    size_t a_stride,
+    const float* restrict w,
+    float* restrict c,
+    size_t cm_stride,
+    size_t cn_stride,
+    const union ${PARAMS} params[restrict XNN_MIN_ELEMENTS(1)])
+{
+  assert(mr != 0);
+  assert(mr <= ${MR});
+  assert(nc != 0);
+  assert(kc != 0);
+  assert(kc % sizeof(float) == 0);
+  assert(a != NULL);
+  assert(w != NULL);
+  assert(c != NULL);
+
+  $if ACTIVATION == "MINMAX":
+    const float vmin = params->scalar.min;
+    const float vmax = params->scalar.max;
+  const float* a0 = a;
+  float* c0 = c;
+  $if MR > 1:
+    const float* a1 = (const float*) ((uintptr_t) a + a_stride);
+    float* c1 = (float*) ((uintptr_t) c + cm_stride);
+    $for M in range(2, MR):
+      const float* a${M} = (const float*) ((uintptr_t) a${M-1} + a_stride);
+      float* c${M} = (float*) ((uintptr_t) c${M-1} + cm_stride);
+
+  const size_t nr = __riscv_vsetvlmax_e32m${LMUL}();
+  size_t vl = nr;
+  if XNN_LIKELY(mr == ${MR}) {
+    do {
+      if XNN_UNLIKELY(nc < nr) {
+        vl = __riscv_vsetvl_e32m${LMUL}(nc);
```

Comment:
this doesnt feel right.  normally we'd just do it unconditionally.
it can also been done unconditionally toward the end of the loop, so the looping overhead/branch can hide any dependency.
If its slow enough to be worth avoiding, consider an NR remainder handler. 
Neon gemm's dont need to know NC until the bottom of the loop where the full MRxNR tile is stored, and then we loop, or a partial tile is saved and we are done.
Low priority as this is outer loop, but if you can find a better way, that would be good.

Reply:
Those conditions inside the loop are to utilize property of vector. They avoid tackling tailed part with extra code like: https://github.com/google/XNNPACK/blob/8e0b5f9ac32a5055f99485e7c92a78f6fc42ace9/src/f32-gemm/gen/f32-gemm-4x4-scalar.c#L140-L163

However, it's worth us to explore which approach is better for performance. SIMD-style or vector-style.